File: lcci/cc/rotate-matrix-lcci.cc
// https://leetcode-cn.com/problems/rotate-matrix-lcci/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        // 水平翻转
        for (int i = 0; i < n / 2; ++i) {
            for (int j = 0; j < n; ++j) {
                swap(matrix[i][j], matrix[n - i - 1][j]);
            }
        }
        // 主对角线翻转
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
    }
};

int main(int argc, char const* argv[]) {
    vector<vector<int>> matrix = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    Solution s;
    s.rotate(matrix);
    for (auto&& arr : matrix) {
        for (auto&& item : arr) {
            cout << item << ",";
        }
        cout << endl;
    }
    return 0;
}

File: lcci/rs/rotate-matrix-lcci.rs
// https://leetcode-cn.com/problems/rotate-matrix-lcci/
#[derive(Debug)]
struct Solution {
}

impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        let n = matrix.len();
        // 水平翻转
        for i in 0..(n / 2) {
            for j in 0..n {
                let tmp = matrix[i][j];
                matrix[i][j] = matrix[n - i - 1][j];
                matrix[n - i - 1][j] = tmp;
            }
        }

        // 主对角线翻转
        for i in 0..n {
            for j in 0..i {
                let tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }
    }
}

fn main() {
    let mut matrix = vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
    Solution::rotate(&mut matrix);
    // println!("{:?}", Solution::rotate(&mut matrix));
    println!("{:?}", matrix);
}
